core/node/node_framework/src/service/mod.rs

```diff
@@ -159,67 +163,128 @@ impl ZkStackService {
         let rt_handle = self.runtime.handle().clone();
         let join_handles: Vec<_> = long_running_tasks
             .into_iter()
-            .map(|task| rt_handle.spawn(task).fuse())
+            .map(|task| {
+                let name = task.id();
+                NamedBoxFuture::new(rt_handle.spawn(task.into_inner()).fuse().boxed(), name)
+            })
             .collect();
 
+        // Collect names for remaining tasks for reporting purposes.
+        let mut tasks_names: Vec<_> = join_handles.iter().map(|task| task.id()).collect();
+
         // Run the tasks until one of them exits.
-        let (resolved, _, remaining) = self
+        let (resolved, resolved_idx, remaining) = self
             .runtime
             .block_on(futures::future::select_all(join_handles));
-        let result = match resolved {
-            Ok(Ok(())) => Ok(()),
-            Ok(Err(err)) => Err(err).context("Task failed"),
+        // Extract the result and report it to logs early, before waiting for any other task to shutdown.
+        // We will also collect the errors from the remaining tasks, hence a vector.
+        let mut errors = Vec::new();
+        let task_name = tasks_names.swap_remove(resolved_idx);
+        match resolved {
```

Comment:
Here and below some refactoring is certainly required, but I would like to postpone it a bit.